### PR TITLE
MAINT: hide symbols on Linux when using GNU linker

### PR DESCRIPTION
**Pull request by pv**
Hide unnecessary symbols in .so libraries, using a GNU linker version script on linux.

This is a general solution to the discussion in https://github.com/scipy/scipy/pull/8451

Overriding the distutils logic by a custom build_ext seemed the cleanest option. The logic of detecting when a GNU linker is similar to what distutils does internally, didn't find a nicer way and this information doesn't seem to be available otherwise.

There was a testing module `scipy.integrate._test_multivariate` that tried to look up some test symbols with ctypes --- rewrote that to use a different approach.

This may also work on OSX, but I'll leave that to someone else.
(Windows already does something like this in standard Python distutils with MSVC, so no changes necessary for that platform.)

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -192,6 +193,48 @@ def run(self):
...
+                    f.write(text)
+                    ext.extra_link_args.append('-Wl,--version-script=' + script_fn)
+
+            old_build_ext.build_extension(self, ext)
+
+        def __is_using_gnu_linker(self, ext):
+            if not sys.platform.startswith('linux'):
+                return False
+
+            # Fortran compilation with gfortran uses it linking. For
```
Comment (tylerjereddy) [inline setup.py]: "its own?"

**Review by tylerjereddy — COMMENTED** [review]
This solution seems more sustainable than a case-by-case enforcement of symbol hiding as in #8451. Perhaps @stuartarchibald can briefly confirm given comments on the previously proposed solution.

I'm guessing we won't have tests to enforce "stable" symbol exposure on different platforms anytime soon--I wonder if that is a good practice or just overkill, in terms of guaranteeing what other libraries have low-level access to, etc.

**Comment by pv** [comment]
Added a test on Travis that checks the number of exported symbols.

**Comment by person142** [comment]
> This may also work on OSX, but I'll leave that to someone else.

I'll note that it doesn't work out-of-the box since the default linker on OSX doesn't support version scripts.

**Review by tylerjereddy — APPROVED** [review]
LGTM & there's a test now in CI for linux symbol exposure

**Comment by stuartarchibald** [comment]
@tylerjereddy I think this will achieve what is desired on Linux with GNU ld.

**Pull request merged**